### ISSUE TITLE
Create visual design for initial admin onboarding flow ("wizard"), to show info about apps that are available to new user accounts

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#2246]
Proposed process.

- @neynah self-assigns this issue.
- This issue gets closed when @kentonv or @jparyani believe we have a visual design that specifies all of the user interactions that need to be specified for the admin onboarding flow.

**Comment by paulproteus** [comment]
Goals/context for admin onboarding & apps available to new user accounts:

- The main purpose of injecting this into the admin onboarding flow is that we need to block the user from creating their first admin account if the apps haven't finished downloading yet. Related to that, if the download has failed, or is taking too long, we should permit the admin to skip past this process, which would indeed result in them not getting any pre-installed apps.

- While we're at it, we could convey information to the user about why these apps are useful. You can see this Windows XP installer video at 7:28 to get a sense of one way (maybe not the best way?) of conveying that the providers of some software believe that some aspect of it is important: https://youtu.be/p8G3yrsy6dY?t=7m28s

- In https://cloud.githubusercontent.com/assets/855341/17003971/606f90a0-4e87-11e6-86e4-dffa6c78b4f9.jpg , this wizard change is summarized as: Upon first install, in the wizard, show info that some apps are preinstalled; no opt-out on this page, **next »** -- however, it due to the possibility that the Sandstorm server is unable to contact the Sandstorm app market, it's important that the UI should allow people to skip this.

- It's not important to us from a product perspective that people be able to customize the default apps collection in great detail here, since they can do it from within the admin interface. Therefore I imagine this initial admin onboarding flow design would look different than the admin interface design, even though they're for the same feature. For example, I imagine there would be no checkboxes in the admin onboarding flow.

Technical facts that relate to the design:

- Within Sandstorm, for app downloads, it's easy to get a progressbar for any one download in progress, which could be styled exactly the same way as the current "Downloading..." progress bar, which happens to use the HTML5 progress bar element: https://css-tricks.com/html5-progress-element/ . It's also technically easy to combine the progress of different downloads, if that's the way you think it makes the most sense as a design.

- We plan to initiate the actual downloading process behind the scenes, as early as possible, so we're hopeful that >80% of users would be 100% downloaded by the time this shows up in the interface.

**Comment by ocdtrekkie** [comment]
@paulproteus Shouldn't a Sandstorm server "just" maintain the latest copy of every default app preloaded, and never download apps on user signup? It would entirely remove the problem of "can't contact the app market".

**Comment by paulproteus** [comment]
@ocdtrekkie A difficulty with that is that if the Sandstorm server is behind a restrictive firewall and it can't contact the app market at _all_, then the admin should be able to succeed at setting up Sandstorm IMHO. There are some alternate implementation strategies -- we bundle the preloaded apps into the "Sandstorm bundle" aka what install.sh downloads, or alternatively install.sh could be the one to download these Sandstorm app packages. But in the (IMHO unlikely) case that something didn't work right in those, it's much easier to communicate the details of a failure when we have a web browser available with the admin (initial admin onboarding aka "wizard"), rather than a text console (install.sh).

**Comment by ocdtrekkie** [comment]
As a note, I totally botched the read on this only being for admin creation. But on that note...

Why is it forcing basic app downloads on an admin on initial setup? Beyond technical failure, admins may just _not want_ to use the pre-install setup, so it should ask, not be like "do it, but have a skip button".

Case in point: I want know the name of person who decided Windows 8 and 10 needs to spend five minutes preinstalling Metro apps for every user who signs in for the first time before they can use the device, particularly when in a business environment, none of these apps get used. Furthermore, as an enterprise user, I want that person found, and shot, for making every enterprise user auto-install Candy Crush Soda Saga.

**Comment by paulproteus** [comment]
Hi @ocdtrekkie,

I would vastly prefer that you not express a wish for violence or death using our issue tracker. If you edit your comment to remove the word "shot" I'll be happy to remove this paragraph from this comment of mine. Consider saying instead that you want the person found, and you given a chance to talk to them.

Having said that:

The Sandstorm code has the chance to do the downloads while the admin interacts with the admin onboarding wizard, long before they get to this question. We can launch downloading in the background when the Sandstorm server starts. so I expect easily >80% of admins will not be slowed down by it. I think it'd be probably near 100%, but I don't have data on this yet, so I decreased my estimate to >80% above. Keep in mind that this uses the server's connectivity, not the client's connectivity, and servers typically have very good connectivity to the Sandstorm app market.

As soon as we have a context in which we can usefully ask if the admin is willing to keep waiting, we will offer them the ability to ignore the download and skip right to user creation.

Is asking admins to wait the main obstacle you see? If so, does the above explanation address your concern?

Let me know!

**Comment by ocdtrekkie** [comment]
Sorry, facetious comment updated acceptably?

**Comment by paulproteus** [comment]
To respond to this specific comment:

> If say... I dunno, Wekan wants to offer a Sandstorm that preinstalls Wekan and nothing else, they shouldn't download an extra hundred megabytes of apps they aren't going to offer.

That makes sense to me! For now this is about the default Sandstorm installation. I think it'd be truly splendid to allow Wekan to provide a hint to the install process that Wekan is the only app that should be pre-installed.

**Comment by ocdtrekkie** [comment]
Data usage and storage is a viable question. Note that, again, using Microsoft as a bad example (I actually like their products mostly, I swear), Microsoft was heavily criticized for background downloading Windows 10 before bothering to ask the user if they wanted it. While for most users, myself included, this wasn't a problem, there were plenty of edge cases.

I would think it'd be preferable to ask people if they want to activate this feature on the onboarding flow, and then start background downloading them then, and presumably by the time the flow is done, they are there.

I would like to think that'd be the ideal way to do all 'default app' installation: You select one to add to the group, and if you check it, you continue to work, and maybe there's a progress bar next to that app that indicates download progress on grabbing it and including it on your server, but it doesn't take you out of using the admin panel.

**Comment by paulproteus** [comment]
> Data usage and storage is a viable question. Note that, again, using Microsoft as a bad example (I actually like their products mostly, I swear), Microsoft was heavily criticized for background downloading Windows 10 before bothering to ask the user if they wanted it. While for most users, myself included, this wasn't a problem, there were plenty of edge cases.

I considered data usage when writing this proposal. Right now it's a 22MB download to get Sandstorm itself.

```bash
paulproteus@slittingmill:/tmp$ wget -q https://dl.sandstorm.io/sandstorm-172.tar.xz ; du -h sandstorm-172.tar.xz
22M	sandstorm-172.tar.xz
```

This is nearly the smallest bit of server software I've ever heard of. I don't think I can run a useful Python web app without download >22MB using `pip`.

That's the status quo. This proposal would add a background download of Rocket.Chat, Davros, Etherpad, and Wekan at initial server start time. That would be 121 megabytes, which is getting into "normal server software size". Here's how I got that number.

```bash
paulproteus@slittingmill:/tmp$ wget -q https://app-index.sandstorm.io/packages/efbc8d5c9586acc919ccbeca1cd18f2e -O wekan.spk ; du -h wekan.spk
31M	wekan.spk
paulproteus@slittingmill:/tmp$ wget -q https://app-index.sandstorm.io/packages/a9fc7e902b320a695a54077a56daf8f4 -O etherpad.spk
paulproteus@slittingmill:/tmp$ du -h etherpad.spk
32M	etherpad.spk
paulproteus@slittingmill:/tmp$ wget -q https://app-index.sandstorm.io/packages/00716d5748099949d0c7f6f285cd1cc2 -O rocket.chat.spk ; du -h rocket.chat.spk
42M	rocket.chat.spk
paulproteus@slittingmill:/tmp$ wget -q https://app-index.sandstorm.io/packages/4fdf2770726db55f2e2f4f0eb96842ab -O davros.spk ; du -h davros.spk
16M	davros.spk
```

One approach we could take is to start downloading 121 megabytes when the user says, "Yes, I want that!" But now the user has to wait for that download to finish.

By contrast, if we begin the download sooner than that, then the download will probably have already finished by the time the user says, "Yes, I want that"!

I believe that 121 megabytes of downloading, from server software, on a server Internet connection, is a pretty OK task to start optimistically if we believe that more than half of our users would want this.

> I would think it'd be preferable to ask people if they want to activate this feature on the onboarding flow, and then start background downloading them then, and presumably by the time the flow is done, they are there.

For what it's worth, you'll see in the mockups, once @neynah prepares them, that this would be during the admin onboarding flow. Perhaps that aspect wasn't clear above; if so, apologies for a lack of clarity on my part.

One alternate proposal that I think I'm reading from you is that, during the admin onboarding flow, we should wait to begin the download until the user agrees. Given that this would slow down most users (since I believe most users would want these 4 apps), and given that to me it's a small amount of data usage on a server, I'm not convinced, but if you can be more specific about what the concerns are, that might help me see things your way.

Perhaps there's another proposal waiting to be found that satisfies both of us more! Or perhaps I've convinced you that the byte counts are low enough and the time savings are worth it to the average admin. Curious for your thoughts.

> I would like to think that'd be the ideal way to do all 'default app' installation: You select one to add to the group, and if you check it, you continue to work, and maybe there's a progress bar next to that app that indicates download progress on grabbing it and including it on your server, but it doesn't take you out of using the admin panel.

This ticket is about the initial admin onboarding flow, aka "wizard". For the proposal for the admin area, see https://github.com/sandstorm-io/sandstorm/issues/2247 & https://github.com/sandstorm-io/sandstorm/issues/2245 . I agree with you that it'd be nice for this post-install admin area functionality to trigger a download in the background, perhaps, though I think there are some challenges worth considering. Let's discuss the details of the post-install admin area on one of those two tickets to avoid confusion. (Note that I just reviewed #2245 and discovered I had linked to the the wrong bug. Perhaps that caused some confusion! If so, apologies.)

**Comment by ocdtrekkie** [comment]
I suppose one could say it's only 120 MB, but you could also make the case it multiplies the size by seven. I'd also highlight that all of these sizes are likely to increase over time, not decrease. At the end of the day it probably doesn't matter too much, but I generally associate unrequested downloads and installs as bad software behavior.

**Comment by kentonv** [comment]
TL;DR

Is this close by #2287?

**Comment by kentonv** [comment]
Yeah, I think this is implemented in #2287. Closing.

**Comment by ocdtrekkie** [comment]
As an addendum to the previous point, I was messing with vagrant-spk today, and it escapes me why my vagrant-spk VM would ever need the pre-installed apps, when really the sole purpose of the thing is to let me try my app in the sandbox.

**Comment by jparyani** [comment]
There was some discussion about making IS_TESTING=true in sandstorm.conf disable preinstalled apps. It would also save our CI from having to constantly download all the packages on every run :)
